### PR TITLE
Lp1850644: Restore slider nature of WOverview

### PR DESCRIPTION
**Pull request by daschuer**
This restores the slider nature of WOverview as requested in 
https://bugs.launchpad.net/mixxx/+bug/1850644

Aborting drag works with the right mouse button. 
Unfortunately the requested ESC key event is not delivered to the widget. 
I have left a comment.

### CODE REVIEW

**Comment by poelzi** [comment]
I tried the ESC key event handler as well and gave up at some point. No idea what eats the event :-/

File: src/widget/woverview.cpp
Comment (ywwg) [inline src/widget/woverview.cpp]: If it doesn't work, let's remove the keyPressEvent handler completely.

File: src/widget/woverview.cpp
Comment (ywwg) [inline src/widget/woverview.cpp]: I don't think we need a TODO for this if right-click works as a cancel method.  But if we do want it, let's say: "TODO: keyPressEvents are not being received by this window during a drag operation. If we want to cancel a drag by ESC we need to figure out why"

**Review by ywwg — CHANGES_REQUESTED** [review]
Right click to cancel was the thing I was going to try -- thanks for doing this.  It works nicely.

Should we keep the keypress code if it doesn't work?  I think it would just add confusion.

**Comment by Be-ing** [comment]
Right clicking to cancel works well with my USB mouse but I cannot use it with my touchpad. Showing the time position on right click should be skipped when cancelling a drag. Touchscreen input is still working fine.

File: src/widget/woverview.cpp
Comment (Be-ing) [inline src/widget/woverview.cpp]: Rename to `m_bLeftClickDragging`

File: src/widget/woverview.h
Comment (Be-ing) [inline src/widget/woverview.h]: move comment one line down

File: src/widget/woverview.cpp
Comment (Be-ing) [inline src/widget/woverview.cpp]: I think this would be better handled in the mouse event code rather than the drawing code. `m_bTimeRulerActive` should not be set to true while dragging.

File: src/widget/woverview.cpp
Comment (Be-ing) [inline src/widget/woverview.cpp]: combine with previous `if` block as `else if (m_pHoveredMark == nullptr) {`

**Review by Be-ing — CHANGES_REQUESTED** [review]

File: src/widget/woverview.cpp
Comment (daschuer) [reply src/widget/woverview.cpp]: It should. That is IMHO a feature. I only messed something up when committing so it was broken. 
Now it is working again.

**Comment by daschuer** [comment]
comments are addressed.

File: src/widget/woverview.cpp
Comment (Be-ing) [inline src/widget/woverview.cpp]: This breaks the time ruler while dragging. The time between the mouse position and the play position does not get updated as the play position moves.

**Comment by Be-ing** [comment]
I find it disorienting that the time ruler is shown while left click is held, but the play position is lost from the overview. I think these should both be shown with separate markers, just like right clicking without left click.

**Comment by ywwg** [comment]
Do we have any test infrastructure for widgets that we could test this by generating fake clicks and drags and checking UI elements / playposition / etc?

**Comment by Be-ing** [comment]
>  Unfortunately the requested ESC key event is not delivered to the widget.

Maybe try `QWidget::grabKeyboard()`?

**Comment by daschuer** [comment]
I don't need the ESC key, since I think the right mouse button release behaves more natural. 

Not showing the original play position during drag was always the case. I actually think it is OK to clearly distinguish between just measuring and dragging so we don't need one.

If this turns out to be an usability issue, we can fix it in a separate PR.

**Comment by daschuer** [comment]
> Do we have any test infrastructure for widgets that we could test this by generating fake clicks and drags and checking UI elements / playposition / etc?

No, we do not have one. Sqish from froglogic works really nice. https://www.froglogic.com/ Maybe they want to support us, I have a contact. We "just" need a contributor with time to evaluate it.

**Comment by Be-ing** [comment]
> I don't need the ESC key, since I think the right mouse button release behaves more natural.

As I said before, this is not accessible on my laptop touchpad. A key press is necessary for the cancellation to be able to be used.

> If this turns out to be an usability issue, we can fix it in a separate PR.

It is a usability issue. I cannot tell where I am seeking from.

**Comment by Be-ing** [comment]
Qt has a [QTest class](https://doc.qt.io/qt-5/qtest.html) for simulating GUI input.

> No, we do not have one. Sqish from froglogic works really nice. https://www.froglogic.com/ Maybe they want to support us, I have a contact. We "just" need a contributor with time to evaluate it.

I am not okay with bringing proprietary dependencies into Mixxx.

**Comment by daschuer** [comment]
This is again ready for merge.

**Comment by Be-ing** [comment]
I disagree that this is ready for merge. I find it confusing to show the time between the dragged-to point and the play position without showing the play position. I think it is also odd that the shaded background is shown up to the dragged position before the mouse button is released. IMO these should be decoupled, just like they are when right clicking without seeking.

**Comment by daschuer** [comment]
Ok, it looks like there is more work to do.
I like to have https://bugs.launchpad.net/mixxx/+bug/1850644 fixed first, and restore the original 2.2 behaviour.
Shall I remove the distance to play position lable for merge? Do you want to take over?

**Comment by ywwg** [comment]
I also can't right click while I'm left clicking but I really don't even see the need for canceling a drag.  We don't have drag cancel for sliders, or knobs, or the waveform, so I don't see why we need it here.

**Comment by ronso0** [comment]
if I'm not mistaking dragging on the waveform should set the play position on button release.
with sliders, dragging affects the parameter instantly

**Comment by daschuer** [comment]
I have found some time: Dragging looks now like  that in Tango: 
grafik

> if I'm not mistaking dragging on the waveform should set the play position on button release.
with sliders, dragging affects the parameter instantly

Yes. 

For me the right click cancel works, but I have never missed it either. For me there is no use case to drag a playing deck. On a paused deck you you have always a second chance to seek.

**Comment by daschuer** [comment]
Is this OK now?

**Comment by Be-ing** [comment]
No, text for all cue labels becomes illegible (semitransparent?) while holding the left mouse button. Also, the appearance of the time ruler line is inconsistent between left clicking and right clicking.

> For me there is no use case to drag a playing deck. On a paused deck you you have always a second chance to seek.

Good point. I think it would be helpful to prevent seeks from WOverview if the deck is playing and the deck fader is all the way down or the crossfader is not all the way to the other side. This would prevent some embarrassing mistakes for me. I thought about making this an option, but do we really need it to be optional?

**Comment by ywwg** [comment]
I don't think we should try to prevent every possible embarrassing mis-input.  We do prevent loading a track into a playing deck (optionally), but there are lots and lots of buttons one can push that could be embarrassing.  (I've been known to accidentally hit CUE at the wrong time). This case would require a lot of complex logic, and then adding another preference for it is more overhead... I don't see a compelling reason to do the extra work.

**Comment by daschuer** [comment]
Ups, now the opacity issue is fixed: 

grafik

The visual difference between right click and left click is intentionally. The right click just watches for time labels and has the small ruler. The left click moves the play head, so the whole play head is moved like on a turn table. This is IMHO required that the user knows what he does.

**Comment by Be-ing** [comment]
> I don't think we should try to prevent every possible embarrassing mis-input. We do prevent loading a track into a playing deck (optionally), but there are lots and lots of buttons one can push that could be embarrassing. (I've been known to accidentally hit CUE at the wrong time). This case would require a lot of complex logic, and then adding another preference for it is more overhead... I don't see a compelling reason to do the extra work.

Of course, we can't prevent the user from making every stupid mistake. I have pressed the wrong play button more times that I would like. That does not mean we should add some safeguard to check the user is pressing the play button they actually intend because there is no way for Mixxx to determine that automatically (however, we should [design a new skin](https://mixxx.zulipchat.com/#narrow/stream/109171-development/topic/design.20brainstorming/near/173236656) so it is much harder to get confused about which play button corresponds to which deck). But we should try to minimize the possibilities for user error so long as it doesn't prevent the user from doing something useful. I can think of no use case for replacing the track in a playing deck; likewise, I can think of no use case for seeking to an arbitrary point in a deck that is presently audible. If a user does actually want to seek to another point in an audible deck, that is what hotcues are for.

The logic for implementing this would not be very complicated. If @daschuer wants to implement it here, great. If not, I'll work on it after this is merged.

**Comment by Be-ing** [comment]
While dragging, the play position line is so thin it is hard to see. I think it should remain the same thickness as usual.
image

> The visual difference between right click and left click is intentionally. The right click just watches for time labels and has the small ruler. The left click moves the play head, so the whole play head is moved like on a turn table. This is IMHO required that the user knows what he does.

Okay, that makes sense. The awkwardness is that they can be different colors. Currently the right click line is hardcoded green. Maybe it should take the play position color from the skin instead.

**Comment by daschuer** [comment]
I have already taken the color from the skin. 
Making the line thin was my intention, because it is only a secondary information and it should not be so prominent. It is more important to be able to look through and see the waveform. We can even considder to dim the cue marker as well.

**Comment by daschuer** [comment]
The original play position should not look as usual because the control is a slider and only if we actually visibility move the slider handle it has the right look and feel. With the current master version I missed it hard to have the natural feel of moving a slider which was my diver here, not gathering info during dragging. Dragging to a new position is usually done quite quickly and the old position is irrelevant IMHO. This PR restores the old behaviour and I have never missed this information.

**Comment by Be-ing** [comment]
> I have already taken the color from the skin.

In the new drawing code you added for left clicking, the color comes from the skin. Please remove this redundant code and use the same drawing code as is used for right clicking, and make that code take the color from the skin. Then there will be no inconsistencies between left and right clicking.

**Comment by Be-ing** [comment]
> Making the line thin was my intention, because it is only a secondary information and it should not be so prominent. It is more important to be able to look through and see the waveform. We can even considder to dim the cue marker as well.

I think using a normal thickness line (the same as the cues) would probably be better, but I don't have a very strong opinion on this so I'm not going to hold up this PR over this issue. What do others think?

>  natural feel of moving a slider

The overview waveform is not a slider. Normally a slider does not instantly jump to a different position when clicked; it has to be dragged. So I consider using the word "slider" in the code misleading.

**Comment by daschuer** [comment]
Done.

**Comment by Be-ing** [comment]
Augh, I noticed after merging that this that #2345 was merged into this branch.

**Pull request merged**
